src/main/java/it/gov/pagopa/cgn/portal/email/EmailNotificationService.java

```diff
@@ -33,22 +41,62 @@ public EmailNotificationService(JavaMailSender javaMailSender) {
+        sendSyncMessage(emailParams, null);
+    }
+
+    public void sendSyncMessage(EmailParams emailParams, String trackingKey) throws MessagingException {
+
+        if (notificationAlreadySent(trackingKey))
+            return;
 
         MimeMessage mimeMessage = javaMailSender.createMimeMessage();
+        try {
+            MimeMessageHelper helper = new MimeMessageHelper(mimeMessage, true, "UTF-8");
+            helper.setFrom(emailParams.getMailFrom());
+            helper.setTo(emailParams.getMailToList().toArray(new String[0]));
+
+            if (emailParams.getReplyToOpt().isPresent()) {
+                helper.setReplyTo(emailParams.getReplyToOpt().orElseThrow());
+            }
+            helper.setSubject(emailParams.getSubject());
+            helper.setText(emailParams.getBody(), true);
+            helper.addInline(emailParams.getLogoName(), emailParams.getLogo());
+
+            log.info("Sending email '{}'", log.isDebugEnabled() ? emailParams.toString() : emailParams.toLightString());
+            javaMailSender.send(mimeMessage);
+            trackNotification(trackingKey);
+        } catch (MessagingException e) {
+            trackNotification(trackingKey, e.getMessage());
+            throw e;
+        }
+    }
+
+    private NotificationEntity findNotification(String trackingKey) {
+        return notificationRepository.findByKey(trackingKey);
+    }
 
-        MimeMessageHelper helper = new MimeMessageHelper(mimeMessage, true, "UTF-8");
-        helper.setFrom(emailParams.getMailFrom());
-        helper.setTo(emailParams.getMailToList().toArray(new String[0]));
+    private void trackNotification(String trackingKey) {
+        trackNotification(trackingKey, null);
+    }
 
-        if (emailParams.getReplyToOpt().isPresent()) {
-            helper.setReplyTo(emailParams.getReplyToOpt().orElseThrow());
+    private void trackNotification(String trackingKey, String errorMessage) {
+        if (trackingKey != null) {
+            var notification = findNotification(trackingKey);
+            if (notification == null) {
+                // create a new Notification
+                notification = new NotificationEntity(trackingKey);
+            }
+            notification.setErrorMessage(errorMessage);
+            notificationRepository.save(notification);
         }
-        helper.setSubject(emailParams.getSubject());
-        helper.setText(emailParams.getBody(), true);
-        helper.addInline(emailParams.getLogoName(), emailParams.getLogo());
+    }
 
-        log.info("Sending email '{}'", log.isDebugEnabled() ? emailParams.toString() : emailParams.toLightString());
-        javaMailSender.send(mimeMessage);
+    private boolean notificationAlreadySent(String trackingKey) {
+        if (trackingKey != null) {
+            var notification = findNotification(trackingKey);
+            return notification != null && notification.getErrorMessage() == null;
```

Comment:
this is not clear. Can you provide a comment?

Reply:
Yes.

Reply:
done https://github.com/pagopa/cgn-onboarding-portal-backend/pull/107/commits/9e9779ea784ca535665776493e2eb2b14a7bb3b3